### PR TITLE
Add dynamically adjustable array to JIT

### PR DESCRIPTION
**Pull request by lyrivera**
## Description

This PR enables the use of an array (named `alpha`) with adjustable parameters within the user defined pair potential.

## Motivation and Context

Allow JIT users to change parameters in their custom pair potential without the need to recompile.

Resolves: No corresponding issue was open

## How Has This Been Tested?

Yes. Unit tests are included.

## Change log

```
 Enable the use of an array with adjustable parameters within the user defined pair potential.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).

### CODE REVIEW

File: hoomd/jit/patch.py [hunk 1]
```diff
@@ -232,6 +268,7 @@ class user_union(user):
         code_iso (str, **optional**): C++ code for isotropic part
         llvm_ir_fname (str): File name of the llvm IR file to load.
         llvm_ir_fname_iso (str, **optional**): File name of the llvm IR file to load for isotropic interaction
+        array_size (int): Size of array with adjustable elements.
```
Comment (joaander) [inline hoomd/jit/patch.py]: - [ ] Note that this is added in version 2.7.

File: hoomd/jit/patch.py [hunk 2]
```diff
@@ -195,6 +198,39 @@ def compile_user(self, code, clang_exec, fn=None):
 
         return llvm_ir
 
+    def set_alpha(self, alpha, index=None):
```
Comment (joaander) [inline hoomd/jit/patch.py]: Could we use an `alpha` property in place of `get_alpha`/`set_alpha`?:
```
>>> patch.alpha = [1,2,3]
>>> print(patch.alpha)
[1,2,3]
```

Comment (lyrivera) [reply hoomd/jit/patch.py]: For sure. Do you want to remove the set/get methods completely?

File: hoomd/hpmc/test-py/enthalpic_interaction.py [hunk 3]
```diff
@@ -198,5 +198,42 @@ def tearDown(self):
...
+                               """
+        system = init.create_lattice(unitcell=lattice.sc(a=2), n=2);
+        self.mc = hpmc.integrate.sphere(seed=1,d=0.1);
+        self.mc.shape_param.set('A',diameter=0);
+
+    def assert_patch(self, patch):
+
+        # check alpha list is set properly
+        patch.alpha = [-1., 2.7, 10]
+        np.testing.assert_almost_equal([-1., 2.7, 10], patch.alpha, decimal=6)
```
Comment (joaander) [inline hoomd/hpmc/test-py/enthalpic_interaction.py]: - [ ] Use of `assert_almost_equal` is discouraged by the numpy documentation, please use `assert_allclose`

File: hoomd/hpmc/test-py/enthalpic_interaction.py [hunk 4]
```diff
@@ -198,5 +198,42 @@ def tearDown(self):
             del self.log;
             context.initialize();
 
+# These tests check the methods for functionality
+class patch_test_alpha_methods(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/test-py/enthalpic_interaction.py]: Makes sense. Will do

File: hoomd/jit/EvalFactory.cc [hunk 5]
```diff
@@ -77,6 +77,20 @@ EvalFactory::EvalFactory(const std::string& llvm_ir)
     #else
     m_eval = (EvalFnPtr) eval.getAddress();
     #endif
+    
+    auto alpha = m_jit->findSymbol("alpha");
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/jit/EvalFactory.cc]: This implementation places `alpha` is a global variable in the user provided code. If the user provides code that does not have `alpha` as as symbol, this error will occur. The user can only do this when passing the LLVM IR file directly, the helper functions that build the code from a snippet add the global array for the user.

File: hoomd/jit/patch.py [hunk 6]
```diff
@@ -121,37 +121,40 @@ def __init__(self, mc, r_cut, code=None, llvm_ir_file=None, clang_exec=None):
...
 
         self.compute_name = "patch"
-        self.cpp_evaluator = _jit.PatchEnergyJIT(hoomd.context.exec_conf, llvm_ir, r_cut);
+        self.cpp_evaluator = _jit.PatchEnergyJIT(hoomd.context.exec_conf, llvm_ir, r_cut, array_size);
         mc.set_PatchEnergyEvaluator(self);
 
         self.mc = mc
         self.enabled = True
         self.log = False
+        self._alpha = [0]*array_size
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/jit/patch.py]: @lyrivera and I discussed this offline. We could use the property to expose alpha as a numpy array that references the memory internally, limiting the need for this additional complexity on the python side and enabling the user to modify individual elements: ``jit.alpha[4] = 1``.

File: hoomd/jit/patch.py [hunk 7]
```diff
@@ -195,6 +198,32 @@ def compile_user(self, code, clang_exec, fn=None):
...
+        Note:
+            Elements of the list **cannot** be set by indexing. Instead, the entire
+            list must be set in a single step.
+
+        """
+        return self._alpha;
+
+    @alpha.setter
+    def alpha(self, alpha):
+        if isinstance(alpha,list) or isinstance(alpha,np.ndarray):
```
[the remark this answers is not in the document]
Comment (vyasr) [reply hoomd/jit/patch.py]: Actually I forgot that you could use `collections.Sequence` in Python2.7. You just need to check which python version it is before importing so you can decide where to import from.

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/jit/patch.py]: *master* no longer supports python 2.7

File: hoomd/jit/EvalFactory.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply hoomd/jit/EvalFactory.cc]: Got it, this is for when you don't call compileUser with code, but using a file. I forgot that was an option, so this makes sense in that context.

File: hoomd/jit/patch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply hoomd/jit/patch.py]: Got it. In that case, we should just use the snippet in my first comment.

File: hoomd/jit/patch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply hoomd/jit/patch.py]: Yes, that API makes the most sense to me. It reduces Python complexity and is the most transparent interface to the underlying C++.

[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/jit/patch.py]: Will switch to that API. Thanks.

**Comment by lyrivera** [comment]
I finally got back to this, apologies for the delay.

File: hoomd/hpmc/test-py/enthalpic_interaction.py [hunk 8]
```diff
@@ -2,7 +2,7 @@
 from __future__ import print_function
 
 import hoomd
-from hoomd import context, data, init, analyze
+from hoomd import context, data, init, analyze, lattice, analyze
```
Comment (vyasr) [inline hoomd/hpmc/test-py/enthalpic_interaction.py]: Importing analyze twice.

**Pull request merged**